File: frontend/lib/api/insights.ts
import { apiClient } from '@/lib/api/client'

export interface IntentCount {
  intent: string
  count: number
}

export interface QuestionTheme {
  question: string
  count: number
}

export interface TrendPoint {
  date: string
  conversations: number
  flagged: number
  escalated: number
}

export interface KnowledgeGapQuestion {
  conversation_id: string
  message_id: string
  question: string
  intent: string | null
  intent_confidence: number | null
  reason: string
  created_at: string
}

export interface InsightsSummary {
  total_conversations: number
  total_messages: number
  user_messages: number
  assistant_messages: number
  escalated_conversations: number
  flagged_messages: number
  low_confidence_messages: number
  negative_feedback_messages: number
  unanswered_messages: number
  escalation_rate: number
  flag_rate: number
}

export interface ConversationInsights {
  period_start: string
  period_end: string
  summary: InsightsSummary
  top_intents: IntentCount[]
  top_questions: QuestionTheme[]
  trend: TrendPoint[]
  knowledge_gaps: KnowledgeGapQuestion[]
}

export async function getInsights(params?: {
  date_from?: string
  date_to?: string
}): Promise<ConversationInsights> {
  const response = await apiClient.get<ConversationInsights>('/api/v1/admin/insights', {
    params,
  })
  return response.data
}

File: backend/app/schemas/insights.py
"""Schemas for conversation mining insights."""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


class IntentCount(BaseModel):
    """A single intent bucket with its message count."""

    intent: str
    count: int


class QuestionTheme(BaseModel):
    """A frequently asked customer question grouped by normalized text."""

    question: str
    count: int


class TrendPoint(BaseModel):
    """Per-day counts for the flag/escalation trend chart."""

    date: str
    conversations: int
    flagged: int
    escalated: int


class KnowledgeGapQuestion(BaseModel):
    """A customer question the bot likely could not answer well."""

    conversation_id: UUID
    message_id: UUID
    question: str
    intent: str | None = None
    intent_confidence: float | None = Field(default=None, ge=0, le=1)
    reason: str
    created_at: datetime


class InsightsSummary(BaseModel):
    """Top-line counts for the mining period."""

    total_conversations: int
    total_messages: int
    user_messages: int
    assistant_messages: int
    escalated_conversations: int
    flagged_messages: int
    low_confidence_messages: int
    negative_feedback_messages: int
    unanswered_messages: int
    escalation_rate: float = Field(ge=0, le=1)
    flag_rate: float = Field(ge=0, le=1)


class ConversationInsights(BaseModel):
    """Aggregated conversation mining insights over a period."""

    period_start: datetime
    period_end: datetime
    summary: InsightsSummary
    top_intents: list[IntentCount] = Field(default_factory=list)
    top_questions: list[QuestionTheme] = Field(default_factory=list)
    trend: list[TrendPoint] = Field(default_factory=list)
    knowledge_gaps: list[KnowledgeGapQuestion] = Field(default_factory=list)
